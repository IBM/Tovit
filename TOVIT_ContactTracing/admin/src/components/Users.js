/* globals alert */
import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { fetchUsers, createUser, deleteUser } from '../helpers/api'

const superPrivilegeLevel = 1
const defaultPrivilegeLevel = 2

const ViewEnum = {
  list: 1,
  new: 2
}

export function Users ({ currentUser }) {
  const [users, setUsers] = useState(undefined)
  const [view, setView] = useState(ViewEnum.list)
  const [newUsername, setNewUsername] = useState('')
  const [newCanManage, setNewCanManage] = useState(false)
  const [createdUsername, setCreatedUsername] = useState('')
  const [createdPassword, setCreatedPassword] = useState('')
  const { t } = useTranslation()
  const usersList = users || []

  const loadUsers = async () => {
    const users = await fetchUsers()
    if (typeof users !== 'undefined') {
      setUsers(users)
    }
  }

  useEffect(() => {
    if (typeof users === 'undefined') {
      loadUsers()
    }
  })

  const onSubmitNewUser = async (event) => {
    event.preventDefault()
    const usernameRegex = /((^[a-z]+)|(^[0-9]+[a-z]+)|(^[a-z]+[0-9]+))+[0-9a-z]+$/i
    if (usernameRegex.test(newUsername)) {
      const privilege = newCanManage ? superPrivilegeLevel : defaultPrivilegeLevel
      const createdUser = await createUser(newUsername, privilege)
      if (typeof createdUser === 'undefined') {
        alert(t('user_create_fail'))
      } else {
        setCreatedUsername(createdUser.username)
        setCreatedPassword(createdUser.password)
        setView(ViewEnum.created)
        loadUsers()
      }
      setNewUsername('')
      setNewCanManage(false)
    } else {
      alert(t('user_create_invalid_username'))
    }
  }

  const onchangeUsername = (event) => {
    setNewUsername(event.target.value)
  }

  const onchangeNewCanManage = () => {
    setNewCanManage(!newCanManage)
  }

  const handleDelete = async (userId, username) => {
    if (window.confirm(`${t('user_delete_confirm')}: ${username}?`)) {
      const deletedUserSuccessfully = await deleteUser(userId)
      if (deletedUserSuccessfully) {
        loadUsers()
      } else {
        alert(t('user_delete_fail'))
      }
    }
  }

  return (
    <div>
      <div class='d-flex justify-content-between flex-wrap flex-md-nowrap align-items-center pt-3 pb-2 mb-3 border-bottom'>
        <h1 class='h2'>Users</h1>
      </div>
      {view === ViewEnum.list && (
        <div>
          <a class='btn btn-dark text-light mb-2' onClick={() => setView(ViewEnum.new)}>
            New user
          </a>
          <ul class='list-group'>
            {usersList.map(user => {
              const canManage = user.privilege === superPrivilegeLevel
              const canDelete = user.id !== currentUser.id
              return (
                <li class='list-group-item user-list'>
                  <strong>{user.username}</strong>
                  {canManage && (
                    <span> (can manage users)</span>
                  )}
                  {canDelete && (
                    <a onClick={() => handleDelete(user.id, user.username)} class='btn btn-danger text-light user-delete'>
                      Delete
                    </a>
                  )}
                </li>
              )
            })}
          </ul>
        </div>
      )}
      {view === ViewEnum.new && (
        <form class='form-signin' onSubmit={onSubmitNewUser}>
          <h2 class='h3 mb-3 font-weight-normal'>{t('user_create_title')}</h2>
          <label for='username' class='sr-only'>{t('user_create_username')}:</label>
          <input value={newUsername} onChange={onchangeUsername} type='text' id='username' name='username' placeholder={t('user_create_username')} class='form-control' />
          <div class='form-check'>
            <input checked={newCanManage} onChange={onchangeNewCanManage} class='form-check-input' type='checkbox' id='can-manage' />
            <label class='form-check-label' for='can-manage'>
              {t('user_create_can_manage_users')}
            </label>
          </div>
          <button class='btn btn-lg btn-warning btn-block mt-3' type='submit'>{t('user_create_submit')}</button>
        </form>
      )}
      {view === ViewEnum.created && (
        <div>
          <h2 class='h3 mb-3 font-weight-normal'>{t('user_create_success')}:</h2>
          <p>{t('user_create_username')}: <strong>{createdUsername}</strong></p>
          <p>{t('user_create_password')}: <strong>{createdPassword}</strong></p>
          <p>{t('user_create_credential_delivery')}</p>
          <a class='btn btn-dark text-light mt-2' onClick={() => setView(ViewEnum.list)}>
            {t('user_create_exit')}
          </a>
        </div>
      )}
    </div>
  )
}
