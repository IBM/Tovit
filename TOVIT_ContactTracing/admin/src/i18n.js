import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import en from './translations/en'

// the translations
// (tip move them in a JSON file and import them)
const resources = {
  en
}

i18n
  .use(LanguageDetector)
  .use(initReactI18next) // passes i18n down to react-i18next
  .init({
    resources,
    fallbackLng: 'en',
    keySeparator: false, // we do not use keys in form messages.welcome
    interpolation: {
      escapeValue: false // react already safes from xss
    },
    detection: {
      order: ['querystring', 'cookie', 'localStorage', 'navigator', 'htmlTag', 'path', 'subdomain'],

      // keys or params to lookup language from
      lookupQuerystring: 'lng',
      lookupCookie: 'i18next',
      lookupLocalStorage: 'i18nextLng',
      lookupFromPathIndex: 0,
      lookupFromSubdomainIndex: 0,

      // cache user language on
      caches: ['localStorage', 'cookie'],
      excludeCacheFor: ['cimode'], // languages to not persist (cookie, localStorage)

      // optional expire and domain for set cookie
      cookieMinutes: 10,
      cookieDomain: 'myDomain',

      // optional htmlTag with lang attribute, the default is:
      htmlTag: document.documentElement,

      // only detect languages that are in the whitelist
      checkWhitelist: true
    }
  })

export default i18n
