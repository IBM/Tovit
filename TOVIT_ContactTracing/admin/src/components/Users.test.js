/* globals test, expect */
import React from 'react'
import { Users } from './Users'
import TestRenderer from 'react-test-renderer'
import { I18nextProvider } from 'react-i18next'
import i18n from '../i18n'

test('Users component', () => {
  const currentUser = {
    username: 'mr_fox',
    privilege: 1,
    id: 'abc123'
  }
  const component = TestRenderer.create(
    <I18nextProvider i18n={i18n}>
      <Users currentUser={currentUser} />
    </I18nextProvider>)
  expect(component).toMatchSnapshot()
})
