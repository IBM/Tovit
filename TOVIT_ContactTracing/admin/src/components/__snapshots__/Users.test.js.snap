// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users component 1`] = `
<div>
  <div
    class="d-flex justify-content-between flex-wrap flex-md-nowrap align-items-center pt-3 pb-2 mb-3 border-bottom"
  >
    <h1
      class="h2"
    >
      Users
    </h1>
  </div>
  <div>
    <a
      class="btn btn-dark text-light mb-2"
      onClick={[Function]}
    >
      New user
    </a>
    <ul
      class="list-group"
    />
  </div>
</div>
`;
